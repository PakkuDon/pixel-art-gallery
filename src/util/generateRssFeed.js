/* eslint-disable no-console */
import fs from "fs"
import path from "path"
import marked from "marked"

import pixelArtEntries from "../data"
import extractFilename from "./extractFilename"

const generateRssFeed = (entries = []) => {
  marked.setOptions({
    breaks: true,
    gfm: true,
    xhtml: true,
  })

  const feedItems = entries.map((pixelArt) => {
    const filename = extractFilename(pixelArt.src)
    const title = pixelArt.description.split("\n")[1].trim()

    return `
      <item>
        <title>${title}</title>
        <description>
          <![CDATA[<img src="https://pakkudon.github.io/pixel-art-gallery/img/${
            pixelArt.src
          }" alt="${filename}" />]]>
          ${marked(pixelArt.description.replace(/ +/g, " "))}
        </description>
        <link>https://pakkudon.github.io/pixel-art-gallery/${filename}</link>
        <pubDate>${new Date(pixelArt.date).toGMTString()}</pubDate>
      </item>
    `
  })

  return `<?xml version="1.0" encoding="ISO-8859-1" ?>
    <rss version="2.0">
      <channel>
        <title>Pixel Art Gallery</title>
        <link>https://pakkudon.github.io/pixel-art-gallery/</link>
        <description>Collection of pixel art pieces</description>
        <pubDate>${new Date().toGMTString()}</pubDate>
        ${feedItems.join("\n")}
      </channel>
    </rss>
  `
}

const outputFile = path.join(__dirname, "../../dist/feed.xml")

if (process.env.NODE_ENV !== "test") {
  const xml = generateRssFeed(pixelArtEntries)
  fs.writeFile(outputFile, xml, (error) => {
    if (error) {
      console.error(error)
    } else {
      console.log(`RSS feed written to ${outputFile}`)
    }
  })
}

export default generateRssFeed
