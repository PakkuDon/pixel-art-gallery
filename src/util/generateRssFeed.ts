/* eslint-disable no-console */
import { marked } from "marked"

import { PixelArtRepository } from "../PixelArtRepository"
import { PixelArtEntry } from "../data"
import { extractFilename } from "./extractFilename"

PixelArtRepository.load()

const generateRssFeed = ({
  entries = [],
  limit,
}: {
  entries: PixelArtEntry[]
  limit?: number
}) => {
  marked.setOptions({
    breaks: true,
    gfm: true,
    xhtml: true,
  })
  let filteredEntries = entries

  if (limit) {
    filteredEntries = entries.slice(-limit)
  }

  const feedItems = filteredEntries.map((pixelArt) => {
    const filename = extractFilename(pixelArt.src)

    return `
      <item>
        <title>${pixelArt.title}</title>
        <description>
          <![CDATA[<img src="https://pakkudon.github.io/pixel-art-gallery/img/${
            pixelArt.src
          }" alt="${filename}" />]]>
          <p>${pixelArt.title || ""}</p>
          ${marked((pixelArt.description || "").replace(/ +/g, " "))}
        </description>
        <link>https://pakkudon.github.io/pixel-art-gallery/${filename}</link>
        <pubDate>${new Date(pixelArt.date).toUTCString()}</pubDate>
      </item>
    `
  })

  return `<?xml version="1.0" encoding="ISO-8859-1" ?>
    <rss version="2.0">
      <channel>
        <title>Pixel Art Gallery</title>
        <link>https://pakkudon.github.io/pixel-art-gallery/</link>
        <description>Pixel art by PakkuDon</description>
        <pubDate>${new Date().toUTCString()}</pubDate>
        ${feedItems.join("\n")}
      </channel>
    </rss>
  `
}

export { generateRssFeed }
