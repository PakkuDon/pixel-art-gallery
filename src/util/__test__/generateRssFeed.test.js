import generateRssFeed from "../generateRssFeed"

describe("generateRssFeed", () => {
  beforeAll(() => {
    jest.useFakeTimers("modern")
    jest.setSystemTime(new Date("2020-01-01").getTime())
  })

  it("generates RSS feed containing pixel art entries", () => {
    const entries = [
      {
        src: "example.png",
        description: `
          Example image with markdown
          Palette: [Example link](https://example.com)
        `,
        date: "2020-01-01 00:00",
        tags: ["a", "b"],
      },
    ]

    expect(generateRssFeed(entries)).toMatchSnapshot()
  })
})
