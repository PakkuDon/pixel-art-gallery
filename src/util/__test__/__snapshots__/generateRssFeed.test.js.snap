// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateRssFeed generates RSS feed containing pixel art entries 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\" ?>
    <rss version=\\"2.0\\">
      <channel>
        <title>Pixel Art Gallery</title>
        <link>https://pakkudon.github.io/pixel-art-gallery/</link>
        <description>Collection of pixel art pieces</description>
        <pubDate>Wed, 01 Jan 2020 00:00:00 GMT</pubDate>
        
      <item>
        <title>Example image with markdown</title>
        <description>
          <![CDATA[<img src=\\"https://pakkudon.github.io/pixel-art-gallery/img/example.png\\" alt=\\"example\\" />]]>
          <p> Example image with markdown<br/> Palette: <a href=\\"https://example.com\\">Example link</a>
 </p>

        </description>
        <link>https://pakkudon.github.io/pixel-art-gallery/example</link>
        <pubDate>Tue, 31 Dec 2019 13:00:00 GMT</pubDate>
      </item>
    
      </channel>
    </rss>
  "
`;

exports[`generateRssFeed when there are more entries than given limit truncates feed to last x items 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"ISO-8859-1\\" ?>
    <rss version=\\"2.0\\">
      <channel>
        <title>Pixel Art Gallery</title>
        <link>https://pakkudon.github.io/pixel-art-gallery/</link>
        <description>Collection of pixel art pieces</description>
        <pubDate>Wed, 01 Jan 2020 00:00:00 GMT</pubDate>
        
      <item>
        <title>Example image with markdown</title>
        <description>
          <![CDATA[<img src=\\"https://pakkudon.github.io/pixel-art-gallery/img/entry-that-will-be-included.png\\" alt=\\"entry-that-will-be-included\\" />]]>
          <p> Example image with markdown<br/> Palette: <a href=\\"https://example.com\\">Example link</a>
 </p>

        </description>
        <link>https://pakkudon.github.io/pixel-art-gallery/entry-that-will-be-included</link>
        <pubDate>Tue, 31 Dec 2019 13:00:00 GMT</pubDate>
      </item>
    
      </channel>
    </rss>
  "
`;
